### PR TITLE
fix: encode basename in restore action

### LINKED ISSUE #39445: [Bug]: "Restore" failed on some elements - String contains non ISO-8859-1 code point
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Hello, I have encountered a bug. After deleting a Chinese name file on the client or web page, I will enter 'deleted files' to restore it, and it will prompt' Restore 'failed on some elements.

The console displays

> [ERROR] files: TypeError: Failed to execute 'setRequestHeader' on 'XMLHttpRequest': String contains non ISO-8859-1 code point

Non Chinese files can be deleted and restored normally.

### Steps to reproduce

1. Create a new Chinese name file such as "Chinese. md"

2. Delete the file

3. Enter deleted files

4. Select the file to restore

### Expected behavior

The file is restored normally.

### Installation method

Other Community project

### Nextcloud Server version

27

### Operating system

Other

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
sudo -u www-data php occ config:list system
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "password": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "customclient_desktop": "https:\/\/n.mmk12.cn:8443\/s\/4z9NQDnkCEkJGL4",
        "customclient_android": "https:\/\/n.mmk12.cn:8443\/s\/kKyTG2p9zk785b4",
        "default_phone_region": "CN",
        "default_language": "zh_CN",
        "default_locale": "zh",
        "force_locale": "zh_CN",
        "auth.bruteforce.protection.enabled": false,
        "simpleSignUpLink.shown": false,
        "skeletondirectory": "",
        "trashbin_retention_obligation": "auto,30",
        "proxy": "192.168.10.3:7890",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "192.168.10.2:1234",
            "n.mmk12.cn"
        ],
        "overwritehost": "n.mmk12.cn:8443",
        "overwriteprotocol": "https",
        "overwrite.cli.url": "http:\/\/192.168.10.2:1234",
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "ssl",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "27.0.0.8",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "app_install_overwrite": [
            "backup"
        ],
        "maintenance": false
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.19.0
  - admin_audit: 1.17.0
  - circles: 27.0.0
  - cloud_federation_api: 1.10.0
  - collectives: 2.6.1
  - comments: 1.17.0
  - contacts: 5.3.2
  - contactsinteraction: 1.8.0
  - dav: 1.27.0
  - federatedfilesharing: 1.17.0
  - federation: 1.17.0
  - files: 1.22.0
  - files_mindmap: 0.0.28
  - files_pdfviewer: 2.8.0
  - files_rightclick: 1.6.0
  - files_sharing: 1.19.0
  - files_trashbin: 1.17.0
  - files_versions: 1.20.0
  - firstrunwizard: 2.16.0
  - groupfolders: 15.0.1
  - logreader: 2.12.0
  - lookup_server_connector: 1.15.0
  - nextcloud_announcements: 1.16.0
  - notifications: 2.15.0
  - notify_push: 0.6.3
  - oauth2: 1.15.0
  - password_policy: 1.17.0
  - photos: 2.3.0
  - privacy: 1.11.0
  - provisioning_api: 1.17.0
  - recommendations: 1.6.0
  - related_resources: 1.2.0
  - richdocuments: 8.1.0
  - serverinfo: 1.17.0
  - settings: 1.9.0
  - sharebymail: 1.17.0
  - support: 1.10.0
  - survey_client: 1.15.0
  - systemtags: 1.17.0
  - text: 3.8.0
  - theming: 2.2.0
  - twofactor_backupcodes: 1.16.0
  - updatenotification: 1.17.0
  - user_status: 1.7.0
  - viewer: 2.1.0
  - weather_status: 1.7.0
  - workflowengine: 2.9.0
Disabled:
  - backup: 1.2.0 (installed 1.2.0)
  - bruteforcesettings: 2.7.0
  - dashboard: 7.7.0 (installed 7.7.0)
  - encryption: 2.15.0
  - files_external: 1.19.0 (installed 1.19.0)
  - files_fulltextsearch: 27.0.0 (installed 27.0.0)
  - fulltextsearch: 27.0.0 (installed 27.0.0)
  - fulltextsearch_elasticsearch: 27.0.0 (installed 27.0.0)
  - spreed: 17.0.1 (installed 17.0.1)
  - suspicious_login: 5.0.0
  - twofactor_totp: 9.0.0
  - user_ldap: 1.17.0
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### LINKED ISSUE #39823: [Bug]: It is impossible to restore a file from the trashbin containing Cyrillic characters in the name
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

It is impossible to restore a file from the trashbin containing Cyrillic characters in the name.
image

In browser console log: *ConsoleLogger.js:59 [ERROR] files: TypeError: Failed to execute 'setRequestHeader' on 'XMLHttpRequest': String contains non ISO-8859-1 code point. {app: 'files', uid: 'admin-test', level: 2}*

### Steps to reproduce

1. Create a file containing Cyrillic characters in the name, for example *мир.md*
2. Delete file
3. Go to the trashbin, try to restore the file.

### Expected behavior

restore the file

### Installation method

Community Docker image

### Nextcloud Server version

27

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.2

### Web server

Nginx

### Database engine version

PostgreSQL

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

Chrome 115.0.5790.171 (latest), Win 11
nextcloud:27.0.2-fpm

### PR DESCRIPTION
**Pull request by kesselb**
* Resolves: #39823 
* Resolves: #39445

## Summary

As per https://xhr.spec.whatwg.org/#the-setrequestheader()-method and https://infra.spec.whatwg.org/#byte-sequences the header value must be in the range 0x20 (SP) to 0x7E (~) hence we need to encode everything else.

## TODO

- [x] CI

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by skjnldsv — APPROVED** [review]
There might be many locations where this would be needed then :thinking:

**Comment by kesselb** [comment]
> There might be many locations where this would be needed then

I guess ;)

File: apps/files_trashbin/src/actions/restoreAction.ts [hunk 1]
```diff
@@ -50,7 +50,7 @@ registerFileAction(new FileAction({
 
 	async exec(node: Node) {
 		try {
-			const destination = generateRemoteUrl(`dav/trashbin/${getCurrentUser()?.uid}/restore/${node.basename}`)
+			const destination = generateRemoteUrl(`dav/trashbin/${getCurrentUser()?.uid}/restore/${encodeURIComponent(node.basename)}`)
```
Comment (skjnldsv) [inline apps/files_trashbin/src/actions/restoreAction.ts:53]:
```suggestion
			const destination = encodeUri(generateRemoteUrl(`dav/trashbin/${getCurrentUser()?.uid}/restore/${node.basename}`))
```

**Review by nickvergessen — APPROVED** [review]

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**